src/sources/nats.rs

```diff
@@ -0,0 +1,182 @@
+use futures::{pin_mut, stream, SinkExt, Stream, StreamExt};
+use serde::{Deserialize, Serialize};
+use snafu::Snafu;
+
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("Could not create Nats subscriber: {}", source))]
+    NatsCreateError { source: std::io::Error },
+    #[snafu(display("Could not subscribe to Nats topics: {}", source))]
+    NatsSubscribeError { source: std::io::Error },
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct NatsSourceConfig {
+    url: String,
+    name: String,
+    subject: String,
+    queue: Option<String>,
+}
+
+inventory::submit! {
+    SourceDescription::new::<NatsSourceConfig>("nats")
+}
+
+impl_generate_config_from_default!(NatsSourceConfig);
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "nats")]
+impl SourceConfig for NatsSourceConfig {
+    async fn build(&self, cx: SourceContext) -> crate::Result<super::Source> {
+        let (connection, subscription) = create_subscription(self).await?;
+
+        Ok(Box::pin(nats_source(
+            connection,
+            subscription,
+            cx.shutdown,
+            cx.out,
+        )))
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        "nats"
+    }
+}
+
+impl NatsSourceConfig {
+    fn to_nats_options(&self) -> async_nats::Options {
+        // Set reconnect_buffer_size on the nats client to 0 bytes so that the
+        // client doesn't buffer internally (to avoid message loss).
+        async_nats::Options::new()
+            .with_name(&self.name)
+            .reconnect_buffer_size(0)
+    }
+
+    async fn connect(&self) -> crate::Result<async_nats::Connection> {
+        self.to_nats_options()
+            .connect(&self.url)
+            .await
+            .map_err(|e| e.into())
+    }
+}
+
+impl From<NatsSourceConfig> for async_nats::Options {
+    fn from(config: NatsSourceConfig) -> Self {
+        async_nats::Options::new()
+            .with_name(&config.name)
+            .reconnect_buffer_size(0)
+    }
+}
+
+fn get_subscription_stream(
+    subscription: async_nats::Subscription,
+) -> impl Stream<Item = async_nats::Message> {
+    stream::unfold(subscription, |subscription| async move {
+        match subscription.next().await {
+            Some(msg) => Some((msg, subscription)),
+            None => None,
+        }
+    })
+}
+
+async fn nats_source(
+    _connection: async_nats::Connection,
+    subscription: async_nats::Subscription,
+    shutdown: ShutdownSignal,
+    mut out: Pipeline,
+) -> Result<(), ()> {
+    let stream = get_subscription_stream(subscription).take_until(shutdown);
+    pin_mut!(stream);
+    while let Some(msg) = stream.next().await {
+        emit!(NatsEventReceived {
+            byte_size: msg.data.len(),
+        });
+
+        let mut event = Event::new_empty_log();
+        let log = event.as_mut_log();
+
+        log.insert(
```

Comment:
What do you think about inserting `timestamp` too that is just "now" to record the time the message was received?